pkg/k8s/apis/cilium.io/utils/utils.go

```diff
@@ -399,5 +399,17 @@ func ParseToCiliumLabels(namespace, name string, uid types.UID, ruleLbs labels.L
 	}
 
 	policyLbls := GetPolicyLabels(namespace, name, uid, resourceType)
-	return append(policyLbls, ruleLbs...).Sort()
+
+	// Ensure user-defined labels from CiliumNetworkPolicy have k8s source
+	userLbls := make(labels.LabelArray, len(ruleLbs))
+	for i, lbl := range ruleLbs {
```

Comment:
Thanks for the fix. I think we should still keep the old fix you had earlier for `UnmarshalJSON` along with this change. https://github.com/cilium/cilium/commit/948be07daf2e20c863fb1e5b4b5ebd173cd60087
When parsing cilium K8s rule we can check for both Upsec and empty source to populate source as K8s here.